### ISSUE TITLE
Training stuck on 50% and prediction script broken.

### ISSUE THREAD

**Issue by Goldie643** [Theano/Theano#5823]
When trying to install cudNN I updated Theano and unfortunately since then my net is stuck on 50% test error (two output channels) even after hundreds of epochs and the method I was using to predict crashes, giving piles of errors in `SeqOptimizer`, primarily

`AttributeError: 'MaxAndArgmax' object has no attribute 'axis'`

and 

`AttributeError: 'AbstractConv2d' object has no attribute 'convdim'`

with the final traceback starting when I define the prediction function

```
predict_model = theano.function(
        inputs=[layer0_input],
        outputs=classifier.y_pred)
```
Both scripts worked perfectly previously. I've tried wiping the OS and reinstalling everything from scratch and yet any version I try and install of any of the libraries I get the same problem. I'm not sure what version of Theano I was running previously (installed everything and didn't update it I believe the start of January, maybe start of December), but I also didn't used to get the prompt to switch to the gpuarray backend so I expect it's a mismatch with the Theano version, though I thought it was entirely backwards compatible.

What's strange is I used the example MNIST net to build mine, only changing the way I loaded the data (from .csv so I can easily inspect the unorthodox data I'm using), and the MNIST example code runs completely fine.

If needed I can post more of the code but given I don't know much about where the problem is stemming I figured I'd see if anyone had any ideas first.

.theanorc file:

```
[global]
mode=FAST_RUN
device=gpu
floatX=float32

[cuda]
root=/usr/local/cuda-8.0

[lib]
cnmem=1
```

**Comment by lamblin** [comment]
Are you reloading graphs or theano functions serialized by Pickle?
If so, that could explain the errors in the optimization phase, which in turn could explain why numerical stability optimization did not get applied, which could prevent your model from converging.

**Comment by Goldie643** [comment]
I'm pickling (cPickle) the model when the validation loss is better than the best via
```
f = open('best_model.pkl', 'wb')
                    cPickle.dump((layer0_input,
                                        layer0,
                                        layer1,
                                        layer2,
                                        layer3_input,
                                        layer3,
                                        classifier),
                                        f)
```
(I know I'm saving more than I need) but in training I'm not reloading a previous model, if that's what you're asking? So it should be completely separate to the training. The loading of the data is just done through csv reader.

The relevant code is [here](https://github.com/Goldie643/cnnt1/blob/master/cnnt_trainer.py) which includes the training and prediction methods. I apologise if it's a bit messy, I pretty much took the MNIST code and fiddled bits here and there.

**Comment by lamblin** [comment]
If you are not reloading previously-pickled model, then I'm not sure where the old versions of `AbstractConv2D` (without `'convdim'`) or `MaxAndArgmax` (without `'axis'`) would come from.

Maybe try `theano-cache purge` and try again?

**Comment by Goldie643** [comment]
When running `theano-cache purge` I get 

```
Error in atexit._run_exitfuncs:
Traceback (most recent call last):
  File "/usr/lib/python2.7/atexit.py", line 24, in _run_exitfuncs
    func(*targs, **kargs)
  File "/usr/local/lib/python2.7/dist-packages/theano/gof/cmodule.py", line 1492, in _on_atexit
    self.clear_old()
  File "/usr/local/lib/python2.7/dist-packages/theano/gof/cmodule.py", line 1295, in clear_old
    cleanup=False)
  File "/usr/local/lib/python2.7/dist-packages/theano/gof/cmodule.py", line 746, in refresh
    subdirs = sorted(os.listdir(self.dirname))
OSError: [Errno 2] No such file or directory: '/home/cnn-tester/.theano/compiledir_Linux-3.19--generic-x86_64-with-Ubuntu-14.04-trusty-x86_64-2.7.6-64'
Error in sys.exitfunc:
Traceback (most recent call last):
  File "/usr/lib/python2.7/atexit.py", line 24, in _run_exitfuncs
    func(*targs, **kargs)
  File "/usr/local/lib/python2.7/dist-packages/theano/gof/cmodule.py", line 1492, in _on_atexit
    self.clear_old()
  File "/usr/local/lib/python2.7/dist-packages/theano/gof/cmodule.py", line 1295, in clear_old
    cleanup=False)
  File "/usr/local/lib/python2.7/dist-packages/theano/gof/cmodule.py", line 746, in refresh
    subdirs = sorted(os.listdir(self.dirname))
OSError: [Errno 2] No such file or directory: '/home/cnn-tester/.theano/compiledir_Linux-3.19--generic-x86_64-with-Ubuntu-14.04-trusty-x86_64-2.7.6-64'
```

Does this mean I've not installed something properly when I reinstalled everything maybe? `/home/cnn-tester/.theano/compiledir_Linux-3.19--generic-x86_64-with-Ubuntu-14.04-trusty-x86_64-2.7.6-64` definitely exists with that path.

**Comment by Goldie643** [comment]
Just to further what you said about loading pre-made models, the prediction script does, could that be why that breaks catastrophically?

**Comment by lamblin** [comment]
Can you remove `/home/cnn-tester/.theano` manually?
I do not understand why things are breaking down like that, if the prediction script loads old model it could give those errors when trying to load them (but it should not have any influence before or after it is run). It might also be possible that loading old models somehow poison the cache.

So the steps would be:
- remove `/home/cnn-tester/.theano` completely
- do not run the prediction script yet
- try to run the training script

**Comment by Goldie643** [comment]
Tried removing .theano and running the training script without touching the prediction script. Still sitting at 50% after 10 epochs, I'll leave it running to see if it begins to converge but it would previously begin dropping after the first epoch. Would it be worth reinstalling Theano and starting off with the training script then? I'm quite stumped with this, the training script should be completely independent. Thank you very much for all your help!

**Comment by lamblin** [comment]
Do you still have the `AttributeError`s at the beginning?
Could you have another copy of your training script (or a file imported by it) somewhere else?

**Comment by Goldie643** [comment]
Do you mean when running the prediction script? I'll check now, the training script doesn't give any errors.

**Comment by Goldie643** [comment]
You were right about the pickle serialisation being the reason for the prediction scripts failure. Using the new model just built the prediction script runs fine, no errors.

Unfortunately, as I was expecting given the strangeness of the net not converging, I've discovered the problem with the training script was me being a little bit of an idiot. When going from training to testing I set my data to not be shuffled on load, and very cleverly forgot to switch it back to shuffling, leaving my data ordered, screwing up the learning. The test error is now decreasing as expected.

Thank you very much for your help! This is for my Masters thesis due in a couple weeks so I'm extremely grateful!
